src/main/java/org/terasology/Lakes/LakeProvider.java

```diff
@@ -15,182 +15,288 @@
+ * If they breach the surface, they're cancelled.
+ * The upper part of their cave is filled with air.
+ * If they're sufficiently far underground, they're lava instead of water.
+ */
 @RegisterPlugin
 @Produces(LakeFacet.class)
+@Updates({
+        @Facet(value = SurfacesFacet.class, border = @FacetBorder(sides = Lake.MAX_RADIUS * 3, bottom = Lake.MAX_DEPTH + 3, top = Lake.MAX_DEPTH + 3)),
+        @Facet(value = DensityFacet.class, border = @FacetBorder(sides = Lake.MAX_RADIUS * 3, bottom = Lake.MAX_DEPTH + 3, top = Lake.MAX_DEPTH + 3))
+})
 @Requires({
-        @Facet(value = SurfaceHeightFacet.class, border = @FacetBorder(sides = (Lake.MAX_SIZE) * 2)),
-        @Facet(value = BiomeFacet.class, border = @FacetBorder(sides = (Lake.MAX_SIZE) * 2))
+    @Facet(value = ElevationFacet.class, border = @FacetBorder(sides = Lake.MAX_RADIUS * 2)),
+    @Facet(value = SeaLevelFacet.class)
 })
 public class LakeProvider implements FacetProviderPlugin {
     private static final int SKIP_BLOCKS = 3;
-    // Don't make any of these constants integer values, the noise functions returns only 0 or 1 for them
-    private static final float SURFACE_LAKES_SAMPLING_CONSTANT = 0.9f;
-    private static final float SURFACE_DESERT_LAKES_SAMPLING_CONSTANT = 0.3f;
-    private static final float UNDERGROUND_LAKES_SAMPLING_CONSTANT = 0.3f;
+    private static final float SURFACE_FREQUENCY = 0.0003f;
+    private static final float SURFACE_EFFECTIVE_FREQUENCY = SURFACE_FREQUENCY * SKIP_BLOCKS * SKIP_BLOCKS * SKIP_BLOCKS;
+    private static final float UNDERGROUND_FREQUENCY = 0.000001f;
+    private static final float UNDERGROUND_EFFECTIVE_FREQUENCY = UNDERGROUND_FREQUENCY * SKIP_BLOCKS * SKIP_BLOCKS * SKIP_BLOCKS;
+    private static final float SURFACE_LAKE_IRREGULARITY = 1.3f;
+    private static final float UNDERGROUND_LAKE_IRREGULARITY = 1.3f;
+
+    Block water;
+    Block lava;
 
-    private Noise noise;
+    private WhiteNoise noise;
+    private Noise depthModifyingNoise;
 
     @Override
     public void setSeed(long seed) {
         // to change the seed value
         noise = new WhiteNoise(seed * 3882);
+        depthModifyingNoise = new SubSampledNoise(new BrownianNoise(new SimplexNoise(seed * 3883), 2), new Vector3f(0.05f, 0.05f, 0.05f), 1);
+        BlockManager blockManager = CoreRegistry.get(BlockManager.class);
+        water = blockManager.getBlock("CoreAssets:Water");
+        lava = blockManager.getBlock("CoreAssets:Lava");
     }
 
     @Override
     public void process(GeneratingRegion region) {
-        SurfaceHeightFacet surfaceHeightFacet = region.getRegionFacet(SurfaceHeightFacet.class);
+        SurfacesFacet surfacesFacet = region.getRegionFacet(SurfacesFacet.class);
+        DensityFacet densityFacet = region.getRegionFacet(DensityFacet.class);
+        ElevationFacet elevationFacet = region.getRegionFacet(ElevationFacet.class);
+        SeaLevelFacet seaLevelFacet = region.getRegionFacet(SeaLevelFacet.class);
```

Comment:
I like how you retrieved all the facets right in the beginning. 

(_just thinking out loud again_)
This looks quite similar to the way we `@ReceiveEvent` and that it would be nice to write 
```java
@ProcessRegion
public void process(GeneratingRegion region, SurfacesFacet surfacesFacet, DensityFacet densityFacet, ...) {
	...
}
```

Reply:
I guess that would be possible. This pattern is very common in facet providers and rasterizers. The `@Requires` and `@Updates` annotations could even be moved here too, to reduce duplication a bit (e.g. something like replacing a `SurfaceHeightFacet` with an `ElevationFacet` would just require one change, rather than three).

Reply:
(_dreamland mode off_)